core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -347,7 +346,7 @@ public JoinReduceExpressionsRule(Class<? extends Join> joinClass,
           matchNullability)) {
         return;
       }
-      if (join instanceof EquiJoin) {
+      if (RelOptUtil.forceEquiJoin(join)) {
```

Comment:
I didn't get this. Can you explain this if block?

Reply:
Cause for current implementation there are some Enumerable joins that do not support non-equi join condition, this is just a sanity check, see @dev discuss: https://lists.apache.org/list.html?dev@calcite.apache.org:lte=1M:support%20non-equi%20join